runtime/compiler/ilgen/Walker.cpp

```diff
@@ -5609,8 +5595,10 @@ break
          ListIterator<TR_VMField> fieldIter(fieldsInfoByIndex->getFields());
          for (TR_VMField *field = fieldIter.getFirst(); field; field = fieldIter.getNext())
             {
-            if ((field->modifiers & J9AccFinal) && !isFinalFieldFromSuperClasses(methodClass, field->offset))
+            if ((field->modifiers & J9AccFinal) && methodClass == jitGetDeclaringClassOfROMField(comp()->j9VMThread(), methodClass, field->shape))
```

Comment:
Would it be worth fetching the declaring class and asserting that it's not NULL (assuming you will never pass a field that's not in the method class hierarchy)?

For the NULL case in this existing code, a fence would not be inserted - is that the correct behaviour? If so, I suppose nothing need be done.

Reply:
wouldn't the conservatively correct thing be to add the fence? I'm not sure what a NULL here would mean from a fence perspective @gita-omr or @zl-wang ?

Reply:
NULL means that the field did not appear in the hierarchy of the provided class (the method class in this case).

Given that the field has been confirmed as final, this is probably a non-issue, as final fields can not be set from outside their defining class in any version of java we support.

Reply:
I would take the NULL answer to mean "the field on hand is not coming from the methodClass".  Thus, no fence is needed.

Reply:
Beyond Java 8, final fields can only be set from within their declaring class.

As an aside, it's not clear to me what final has to do with fences - afaik, all objects require a store fence after construction (complete construction - there's no need for the intermediate fences for super constructors if you're inlining them) , and any fences in the middle of construction are likely unncessary,

Reply:
Incorrect.   The fence you mentioned is on NewObject (unconditionally).   This fence is inserted conditionally on <init>.   Escape analysis can collapse the two into one.

Reply:
The prior one is to safeguard the visibility of object-header.  you need it always.

Reply:
By and large,  the object will not escape during </init/>.  Hope EA would do a good job of collapsing them.

Reply:
> Given that the field has been confirmed as final, this is probably a non-issue, as final fields can not be set from outside their defining class in any version of java we support.

@gacholio Sounds like this should be left as is then. Do you still have concerns about this?